### ISSUE TITLE
M3u Playlists have empty track names

### ISSUE THREAD

**Issue by Joolee** [mopidy/mopidy#1364]
I noticed that most of my playlists show empty track names when viewed through any of the web clients.
After a bit of digging, it seems while indexing playlist, track length and name are only fetched when using extended M3U playlists. (translator.py~102-111) This could be fixed in a number of ways;
* Use Gstreamer to get file information when track.name and/or track.length are empty
* Use at least the filename as source for track.name (extremely easy fix*, though, still no track length)

My favourite but will take the most work:
* Integrate with library plugin. Match tracks with the local database whenever possible. Would solve this and would also add support for extra track information such as album art.
  * Easier; only look up track name and length in local database whenever possible. Store tracks as done now.

The M3U plugin currently seems to be undervalued by the team and desperately needs some love. There are other problems, such as lack of support for UTF8 BOM and lack of caching (starting Mopidy takes +-40 seconds longer for me with playlists enabled). Shall I make Issues for these also?

<span>* Change translator.py line 107 to</span>
```
            name = os.path.basename(line)
            tracks.append(track.replace(uri=uri,name=name))
```

**Comment by adamcik** [comment]
I think we used to have a bug for some of this, but I think we closed it when we did some API changes related to this. Basically the way I think this is supposed to work is that playlist backend providers should be returning just `Ref` models which only contain a name, type and an URI. This way frontend etc could then use library `lookup` calls to get the rest of the metadata. Then any file uris that get looked up would be checked by the new file backend and you should get your metadata.

It seems we might have left some of this hanging half done, or at the very least not done a good enough job getting the web clients on board with using the newer APIs.

The old scheme was just returning the `Track` models which meant a bunch of problems with not having the data we need to fill the metadata without having a playlist backend query the core library. But we don't want backends talking to core in fear of deadlocks, so we never went that route.

As for your favorite variant, you end up with either just supporting the library types you integrate with, and M3U playlists should support any URIs, or you have the same problem as we were trying to avoid where backends need to talk to backends and you risk deadlocks.

As to the M3U extension not getting enough love I do agree. Part of this is just time and focusing more on core bits, and a tentative plan start using Totem's plparser libraries once our switch to GStreamer 1.0 is done. If you have time and interest to take on brushing this extension this would be great. Go ahead and file bugs for any other things you think should be addressed and we can figure out where to go with all this :-)

**Comment by kingosticks** [comment]
I can tell you that musicbox-webclient still doesn't use the API properly (I.e does not do lookups on the returned refs) and gets this all wrong. 
On 20 Dec 2015 13:06, Thomas Adamcik <notifications@github.com> wrote:I think we used to have a bug for some of this, but I think we closed it when we did some API changes related to this. Basically the way I think this is supposed to work is that playlist backend providers should be returning just Ref models which only contain a name, type and an URI. This way frontend etc could then use library lookup calls to get the rest of the metadata. Then any file uris that get looked up would be checked by the new file backend and you should get your metadata.

It seems we might have left some of this hanging half done, or at the very least not done a good enough job getting the web clients on board with using the newer APIs.

The old scheme was just returning the Track models which meant a bunch of problems with not having the data we need to fill the metadata without having a playlist backend query the core library. But we don't want backends talking to core in fear of deadlocks, so we never went that route.

As for your favorite variant, you end up with either just supporting the library types you integrate with, and M3U playlists should support any URIs, or you have the same problem as we were trying to avoid where backends need to talk to backends and you risk deadlocks.

As to the M3U extension not getting enough love I do agree. Part of this is just time and focusing more on core bits, and a tentative plan start using Totem's plparser libraries once our switch to GStreamer 1.0 is done. If you have time and interest to take on brushing this extension this would be great. Go ahead and file bugs for any other things you think should be addressed and we can figure out where to go with all this :-)

**Comment by Joolee** [comment]
I have tried modifying my playlists to use the local:track:relative/from/media/dir.mp3 format but that throws other errors:
````
[...]
Dec 20 14:29:48 raspberrypi mopidy[4887]: ERROR    Got un-handled exception from M3UBackend
Dec 20 14:29:48 raspberrypi mopidy[4887]: Traceback (most recent call last):
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/commands.py", line 239, in _actor_error_handling
Dec 20 14:29:48 raspberrypi mopidy[4887]: yield
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/commands.py", line 380, in start_backends
Dec 20 14:29:48 raspberrypi mopidy[4887]: config=config, audio=audio).proxy()
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/pykka/actor.py", line 94, in start
Dec 20 14:29:48 raspberrypi mopidy[4887]: obj = cls(*args, **kwargs)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/m3u/actor.py", line 35, in __init__
Dec 20 14:29:48 raspberrypi mopidy[4887]: self.playlists = M3UPlaylistsProvider(backend=self)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/m3u/playlists.py", line 28, in __init__
Dec 20 14:29:48 raspberrypi mopidy[4887]: self.refresh()
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/m3u/playlists.py", line 72, in refresh
Dec 20 14:29:48 raspberrypi mopidy[4887]: tracks = translator.parse_m3u(path)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/m3u/translator.py", line 112, in parse_m3u
Dec 20 14:29:48 raspberrypi mopidy[4887]: tracks.append(track.replace(uri=line))
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/models/immutable.py", line 214, in replace
Dec 20 14:29:48 raspberrypi mopidy[4887]: other = super(ValidatedImmutableObject, self).replace(**kwargs)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/models/immutable.py", line 116, in replace
Dec 20 14:29:48 raspberrypi mopidy[4887]: other._set_field(key, value)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/models/immutable.py", line 186, in _set_field
Dec 20 14:29:48 raspberrypi mopidy[4887]: object.__setattr__(self, name, value)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/models/fields.py", line 48, in __set__
Dec 20 14:29:48 raspberrypi mopidy[4887]: value = self.validate(value)
Dec 20 14:29:48 raspberrypi mopidy[4887]: File "/usr/lib/python2.7/dist-packages/mopidy/models/fields.py", line 96, in validate
Dec 20 14:29:48 raspberrypi mopidy[4887]: return intern(str(super(Identifier, self).validate(value)))
Dec 20 14:29:48 raspberrypi mopidy[4887]: UnicodeEncodeError: 'ascii' codec can't encode characters in position 103-105: ordinal not in range(128)
````
They are perhaps not that difficult to fix but that goes a bit over my head.

I could try my hand on the UTF8 fix though.

**Comment by tkem** [comment]
@Joolee: For `#EXTINF` track names to be interpreted as UTF-8, you should use the `.m3u8` file extension with Mopidy v1.1. `.m3u` files will always be read as Latin-1, see also https://en.wikipedia.org/wiki/M3U#M3U8.

As for parsing and setting the track length, this has been kept mostly for backwards compatibility; future versions of Mopidy will no longer return `Track` objects for playlist items, but return `Ref`s instead, as already done by [`PlaylistsProvider.get_items()`](http://mopidy.readthedocs.org/en/latest/api/backend/#mopidy.backend.PlaylistsProvider.get_items). But as @kingosticks already mentioned, not all clients have switched to the new Mopidy v1.0 API methods yet.

**Comment by Joolee** [comment]
Please ignore my previous comment. That was due to a lack of urlencode on my playlist.

Following patch adds filename (minus file extension) to track.name for non-extended m3u playlists. It also handles UTF8 files more gracefully.

````diff
3a4
> import chardet
78c79,89
<     file_encoding = 'utf-8' if file_path.endswith(b'.m3u8') else 'latin1'
---
>
>     bytes = min(32, os.path.getsize(file_path))
>     raw = open(file_path, 'rb').read(bytes)
>     if raw.startswith(codecs.BOM_UTF8):
>         file_encoding = 'utf-8-sig'
>     elif file_path.endswith(b'.m3u8'):
>       file_encoding = 'utf-8'
>     else:
>         result = chardet.detect(raw)
>         file_encoding = result['encoding']
>
83a95
>
101a114,118
>
>         if not track.name:
>             name=os.path.basename(line)
>             name=urllib.parse.unquote(name)
>             track=track.replace(name=os.path.splitext(name)[0])
````

@tkem Musicbee is a player that exports UTF8 m3u playlists as .m3u with BOM. But I can imagine there are other programs that export even .m3u8 playlists with BOM. This change should support both.
As for the track name, I would have written these changes for myself anyway so feel free to implement it otherwise. I look forward to the update.

**Comment by tkem** [comment]
@Joolee: Sorry, I didn't quite understand what you were up to ;-)

IMHO, BOM support as shown in your patch would be a useful feature, since I expect quite a few (Windows) players to include that in exported files, due to the limitations of the M3U format. I also think it shouldn't have too much impact performance-wise. I'm still puzzled by your 40+ seconds startup time increase - how many playlists/tracks are we talking about?

As for falling back on the file path for missing track names, I could swear this was already implemented the way you suggest... But apparently, memory fails me, so I would also welcome that.

**Comment by adamcik** [comment]
The startup time part is probably because we have `@property` in playlists, so pykka's proxy creation triggers loading of that attribute... So just faking it with `return []` in [core/playlists.py line 97](https://github.com/mopidy/mopidy/blob/develop/mopidy/core/playlists.py#L97) would probably "fix" the slowdown. @jodal has worked on this before to speedup mopidy-spotify.

**Comment by tkem** [comment]
@Joolee: Would you mind creating a proper PR for the "fallback on file name" part? I think that would be a useful contribution to the M3U backend without too many side effects.
Regarding the BOM part, this would also affect the "write" part of the playlists API, and I guess we'll have to figure out how to deal with this in general, so I'd postpone this for a while.

**Comment by Joolee** [comment]
Me making a prober PR would require finally understanding GIT. Sadly, I have tried that a few times without success.

Regarding the BOM issue, if I read the code right. The files are currently always written with latin encoding. This while reading the files _already_ has UTF-8 support. I don't see any problem with improving the reading part without worsening the already broken write part.

I could try improving the write part as; _when destination exists, read the encoding of the current file and use the same settings for writing._ Maybe also try to detect non-ascii characters in the to-be saved tracknames when destination doesn't exist yet but this would probably break url's when switching to .m3u8 file extension or break external programs when suddenly introducing a non-standard BOM in the file.

**Comment by tkem** [comment]
@Joolee: That's fine, you already helped a lot by bringing this up and providing your patches. I'll try and give the M3U backend some love myself after the holidays.

**Comment by tkem** [comment]
@Joolee: I opened a new issue for the UTF-8 issues, since that makes it easier to track these separately; especially given this issue's title ;-)
Please feel free to comment on #1370, too.

**Comment by jodal** [comment]
Fixed with the merge of PR #1369.
